File: src/cli/resolve.ts
/**
 * `ocx resolve` — the machine-readable runtime resolution surface for an embedding shell.
 *
 * D5 of devlog/_plan/260921_app_runtime_ownership/: the desktop shell must stop resolving
 * the config home, the port and liveness itself. The tuned probe budgets in
 * src/server/proxy-liveness.ts exist because a shell-side reimplementation answered
 * "nobody listening" twice and started duplicate proxies; this verb exposes that module's
 * verdict instead of copying it, alongside src/config/paths.ts (the home) and the CLI's
 * own preferred-port selection (`config.port ?? 10100` — resolve takes no --port).
 *
 * Contract:
 *  - `--json` puts exactly ONE JSON document on stdout, versioned by `schema`; the
 *    default prints two human lines, the same opt-in split as `ocx ready --json`.
 *  - liveness has three answers, not two: "live", "absent-proven" (every recorded and
 *    configured endpoint definitively refused or answered non-opencodex), and unknown.
 *    Unknown NEVER reaches the wire as absent — a probe that timed out, a listener that
 *    withheld /healthz, or an identity mismatch exits 1 instead. Only "absent-proven"
 *    may authorise starting a new runtime.
 *  - exit 0 whenever a trustworthy verdict exists — live, or proven absent. A MISSING
 *    config.json is defaults, not an error.
 *  - exit 1 when the CLI cannot resolve: an invalid config.json must NOT be answered
 *    with `loadConfig`'s repair-to-defaults behaviour, because that hands the caller
 *    a guessed port; and unknown liveness must not be answered as absence.
 *  - exit 64 for any argument, pre-parsed in src/cli/root.ts before preflight side
 *    effects, mirroring `ocx ready`. The verb is read-only and listed in
 *    skipsCodexShimAutoRestore, so a lookup made to populate a consent surface never
 *    triggers a shim repair side effect.
 *
 * Discovery uses the START_OWNERSHIP_LIVENESS budget, not the 750ms single-shot default:
 * the shell's launch decision keys on this verdict, and answering "nobody" for a slow
 * live proxy is the duplicate-proxy decision the start path tunes against (#5004).
 *
 * Lives outside cli/index.ts (which dispatches argv at module top level) so tests can
 * import it, the same split as ready.ts.
 */
import { readConfigDiagnostics, type ConfigDiagnostics } from "../config";
import { getConfigDir } from "../config/paths";
import { readRuntimePort } from "../config/process-state";
import { packageVersion } from "../lib/package-version";
import {
  findLiveProxy,
  probeEndpointLiveness,
  START_OWNERSHIP_LIVENESS,
  type EndpointLiveness,
  type LiveProxy,
} from "../server/proxy-liveness";
import { endpointsToProve, everyEndpointProvenDownAsync, type ProbeEndpoint } from "./uninstall-plan";
import {
  resolveServiceOwnership,
  resolveServiceState,
  type ServiceInstallState,
  type ServiceOwnershipResolution,
  type ServiceStateResolution,
} from "../service/state";
import {
  assessServiceTakeoverCompatibility,
  type ManagingCliObservation,
  type ManagingCliRole,
} from "../service/ownership-compatibility";
import { observeManagingClis } from "../service/managing-cli";
import { SERVICE_OWNERSHIP_MINIMUM_CLI_VERSION } from "../service/install-state-contract.mjs";

/** Wire version of the resolve document. Bump only on an incompatible shape change. */
export const RESOLVE_SCHEMA = "ocx-resolve/1";

/** The port every preferred-port selection in the CLI falls back to. */
export const RESOLVE_DEFAULT_PORT = 10100;

export interface ResolveLivenessJson {
  /**
   * "live" when the identity-checked probe found our proxy; "absent-proven" when every
   * recorded and configured endpoint is definitively dead. The third state — unknown —
   * exits 1 before this document is printed, so it never appears on the wire as absence.
   */
  status: "live" | "absent-proven";
  pid: number | null;
  port: number | null;
  /** Raw bind hostname that answered; compose probe URLs via probeHostname semantics. */
  hostname?: string;
  /** Where the verdict came from: the runtime record, or the configured listen port. */
  source: LiveProxy["source"] | null;
  /** Version the live proxy reported on /healthz, when it reported one. */
  version?: string;
  /** Listener role the live proxy reported, when it reported one ("client" = connected client). */
  role?: string;
}

export interface ResolveJson {
  schema: typeof RESOLVE_SCHEMA;
  /** Version of this CLI binary, so a shell can compare its engine against the live proxy. */
  cliVersion: string;
  /** Resolved opencodex home (OPENCODEX_HOME or ~/.opencodex), from src/config/paths.ts. */
  configHome: string;
  port: {
    /** The port a client should use: the live listener's port when one answers, else the configured one. */
    effective: number;
    /** The configured listen port (config.port ?? 10100); what a start would prefer. */
    configured: number;
    /** Whether `effective` came from a live proxy or from configuration. */
    source: LiveProxy["source"];
  };
  liveness: ResolveLivenessJson;
  /**
   * The recorded runtime owner, in the CLI's own three answers. `unknown` is on the wire
   * deliberately: it never changes the exit code — the liveness verdict is still
   * trustworthy — and the embedding shell fails closed on it rather than asking consent
   * against a record it could not read.
   */
  ownership: ServiceOwnershipResolution;
  /**
   * Whether a desktop takeover can be offered, and the token that binds that approval to
   * the exact subject and managing-CLI observations a later `ocx service claim` must find
   * unchanged. `ownership-unknown` is produced only here: it is a wire reason, not a new
   * member of `ServiceTakeoverCompatibility`'s union.
   */
  takeover: ResolveTakeover;
}

export type ResolveTakeover =
  | { kind: "supported"; protocolVersion: number; minimumCliVersion: string; token: string }
  | { kind: "blocked"; reason: string; detail: string; minimumCliVersion: string };

export interface ResolveArgs {
  json: boolean;
}

export type ResolveParseResult = { ok: true; args: ResolveArgs } | { ok: false; code: 64 };

/** Pure argument parser: the only flag is `--json`. */
export function parseResolveArgs(argv: string[]): ResolveParseResult {
  for (const flag of argv) {
    if (flag !== "--json") return { ok: false, code: 64 };
  }
  return { ok: true, args: { json: argv.includes("--json") } };
}

export interface ResolveIo {
  configDir?: () => string;
  readDiagnostics?: () => ConfigDiagnostics;
  findLive?: () => Promise<LiveProxy | null>;
  /** Runtime-port record reader; production default is readRuntimePort. */
  readRuntime?: () => { port?: number; hostname?: string } | null;
  /** Tri-state endpoint probe; production default runs in-process for compiled standalone binaries. */
  probeEndpoint?: (endpoint: ProbeEndpoint) => EndpointLiveness | Promise<EndpointLiveness>;
  cliVersion?: () => string;
  /** Recorded-ownership resolver; production default is resolveServiceOwnership. */
  resolveOwnership?: () => ServiceOwnershipResolution;
  /** Full install-state resolver; production default is resolveServiceState. */
  resolveState?: () => ServiceStateResolution;
  /** Managing-CLI observer; production default is observeManagingClis. */
  observeManagers?: (
    state: ServiceInstallState | null,
  ) => Readonly<Record<ManagingCliRole, ManagingCliObservation>>;
  stdout?: { log: (s: string) => void };
  stderr?: { error: (s: string) => void };
}

function livenessJson(live: LiveProxy | null): ResolveLivenessJson {
  // Reaching here with null means absence was PROVEN by the caller (unknown exits 1
  // before this document is built).
  if (!live) return { status: "absent-proven", pid: null, port: null, source: null };
  return {
    status: "live",
    pid: live.pid,
    port: live.port,
    source: live.source,
    ...(live.hostname === undefined ? {} : { hostname: live.hostname }),
    ...(live.version === undefined ? {} : { version: live.version }),
    ...(live.role === undefined ? {} : { role: live.role }),
  };
}

/** Pure shaper: one live verdict plus configuration becomes the wire document. */
export function buildResolveJson(
  config: { port?: number },
  live: LiveProxy | null,
  configHome: string,
  cliVersion: string,
  ownership: ServiceOwnershipResolution,
  takeover: ResolveTakeover,
): ResolveJson {
  const configured = config.port ?? RESOLVE_DEFAULT_PORT;
  return {
    schema: RESOLVE_SCHEMA,
    cliVersion,
    configHome,
    port: {
      effective: live ? live.port : configured,
      configured,
      source: live ? live.source : "config",
    },
    liveness: livenessJson(live),
    ownership,
    takeover,
  };
}

/**
 * Human form: two lines, no prose flourish — an operator skims it, a shell uses --json.
 */
function reportHuman(json: ResolveJson, stdout: { log: (s: string) => void }): void {
  stdout.log(`Config home: ${json.configHome}`);
  const live = json.liveness;
  if (live.status === "live") {
    const pidText = live.pid === null ? "unknown" : String(live.pid);
    const versionText = live.version ?? "unknown version";
    stdout.log(`Proxy live on port ${json.port.effective} (PID ${pidText}, ${versionText}); effective port ${json.port.effective}.`);
  } else {
    stdout.log(`No live proxy (absence proven); effective port ${json.port.effective} (configured).`);
  }
  const ownership = json.ownership;
  if (ownership.kind === "owned") {
    stdout.log(`Owner: ${ownership.ownership.owner} (install ${ownership.ownership.installId}, generation ${ownership.ownership.consentGeneration})`);
  } else if (ownership.kind === "unknown") {
    stdout.log(`Owner: unknown (${ownership.reason})`);
  } else {
    stdout.log("Owner: none recorded");
  }
  stdout.log(
    json.takeover.kind === "supported"
      ? "Takeover: supported"
      : `Takeover: blocked (${json.takeover.reason}: ${json.takeover.detail})`,
  );
}

/**
 * Run `ocx resolve` over injected I/O. Returns the exit code. The production defaults
 * read config through the diagnostics surface (which distinguishes missing, valid and
 * invalid instead of repairing to defaults) and perform one identity-checked discovery
 * at the ownership-safe budget — resolve adds no probing policy of its own.
 */
export async function runResolve(args: ResolveArgs, io: ResolveIo = {}): Promise<number> {
  const stdout = io.stdout ?? console;
  const stderr = io.stderr ?? console;
  const configDir = io.configDir ?? getConfigDir;
  const readDiagnostics = io.readDiagnostics ?? readConfigDiagnostics;
  const findLive = io.findLive ?? (() => findLiveProxy(START_OWNERSHIP_LIVENESS));
  const readRuntime = io.readRuntime ?? readRuntimePort;
  const probeEndpoint = io.probeEndpoint ?? probeEndpointLiveness;
  const cliVersion = io.cliVersion ?? packageVersion;
  const resolveOwnership = io.resolveOwnership ?? resolveServiceOwnership;
  const resolveState = io.resolveState ?? resolveServiceState;
  const observeManagers = io.observeManagers ?? observeManagingClis;
  const configHome = configDir();
  let diagnostics: ConfigDiagnostics;
  try {
    diagnostics = readDiagnostics();
  } catch (error) {
    // A resolution that could not run must not read as "no proxy": the caller has to
    // refuse to guess (D5) rather than treat this as a proven-absent verdict.
    stderr.error(`resolve failed: ${error instanceof Error ? error.message : String(error)}`);
    return 1;
  }
  if (diagnostics.source === "fallback") {
    // An invalid config must not resolve to defaults: the effective port would be a
    // guess at 10100 while the operator's config.port is unread. The repair-to-defaults
    // policy in loadConfig is for interactive recovery, not for a shell contract.
    stderr.error(`resolve failed: the config in ${configHome} is invalid (${diagnostics.error ?? "unknown error"}); refusing to guess.`);
    return 1;
  }
  let live: LiveProxy | null;
  try {
    live = await findLive();
  } catch (error) {
    stderr.error(`resolve failed: ${error instanceof Error ? error.message : String(error)}`);
    return 1;
  }
  if (!live) {
    // findLiveProxy collapses "definitely nothing" and "could not tell" into the same
    // null. The launch decision keys on this verdict, so resolve owes the caller the
    // tri-state answer the updater already enforces: only EVERY candidate definitively
    // dead is absence. Anything else is unknown, and unknown exits 1 — it must never
    // authorise starting a second runtime.
    let provenDown = false;
    try {
      provenDown = await everyEndpointProvenDownAsync(endpointsToProve(readRuntime(), diagnostics.config), probeEndpoint);
    } catch {
      // A probe that cannot run is not evidence of absence.
      provenDown = false;
    }
    if (!provenDown) {
      stderr.error("resolve: liveness is unknown (a probe timed out or a listener withheld /healthz); refusing to treat unknown as absent.");
      return 1;
    }
  }
  let ownership: ServiceOwnershipResolution;
  try {
    ownership = resolveOwnership();
  } catch (error) {
    ownership = { kind: "unknown", reason: error instanceof Error ? error.message : String(error) };
  }
  let takeover: ResolveTakeover;
  if (!live) {
    // A proven absence may authorize start, but there is no runtime to take over.
    // Avoid synchronous managing-CLI version probes on this launch path.
    takeover = {
      kind: "blocked",
      reason: "runtime-absent",
      detail: "no live runtime is available for takeover",
      minimumCliVersion: SERVICE_OWNERSHIP_MINIMUM_CLI_VERSION,
    };
  } else if (ownership.kind === "unknown") {
    // The claim cannot be read, so nothing can be approved against it. This reason is a
    // wire answer, not a new member of the compatibility union.
    takeover = {
      kind: "blocked",
      reason: "ownership-unknown",
      detail: ownership.reason,
      minimumCliVersion: SERVICE_OWNERSHIP_MINIMUM_CLI_VERSION,
    };
  } else {
    try {
      const resolved = resolveState();
      if (resolved.kind === "unknown") {
        takeover = {
          kind: "blocked",
          reason: "ownership-unknown",
          detail: resolved.reason,
          minimumCliVersion: SERVICE_OWNERSHIP_MINIMUM_CLI_VERSION,
        };
      } else {
        const resolvedState = resolved.kind === "state" ? resolved.state : null;
        takeover = assessServiceTakeoverCompatibility({
          state: resolvedState,
          subject: ownership,
          managers: observeManagers(resolvedState),
        });
      }
    } catch (error) {
      takeover = {
        kind: "blocked",
        reason: "managing-cli-unknown",
        detail: error instanceof Error ? error.message : String(error),
        minimumCliVersion: SERVICE_OWNERSHIP_MINIMUM_CLI_VERSION,
      };
    }
  }
  const json = buildResolveJson(diagnostics.config, live, configHome, cliVersion(), ownership, takeover);
  if (args.json) stdout.log(JSON.stringify(json));
  else reportHuman(json, stdout);
  return 0;
}

File: desktop/src-tauri/src/resolve.rs
//! What the bundled CLI says about this machine's runtime.
//!
//! D5: the shell stops resolving the configuration home, the port and liveness itself. It used to,
//! in a file called `discovery.rs` that read `runtime-port.json`, fell back to 10100 and started on
//! that port — so a user with a configured `config.port` was started somewhere else. The tuned probe
//! budgets it should have been using exist because a shell-side reimplementation answered "nobody is
//! listening" twice and started duplicate proxies. This asks instead.
//!
//! Liveness has three answers and the third one is the point. `live` means attach. `absent-proven`
//! means every recorded and configured endpoint was definitively dead, and only that authorises
//! starting a runtime. Anything else is unknown, and the CLI exits 1 rather than putting absence on
//! the wire. Everything that can go wrong on this side — a missing binary, a timeout, output that
//! will not parse, a schema this shell does not know — folds into the same unknown, because the one
//! reading that must never happen is "the resolve failed, so nobody must be listening".

use crate::endpoint::ProxyEndpoint;
use crate::ownership::Recorded;
use serde::Deserialize;
use std::path::PathBuf;
use tauri::AppHandle;
use tauri_plugin_shell::ShellExt;
use tokio::time::{timeout_at, Instant};

/// The wire version this shell understands. A document announcing anything else is unknown.
pub const SCHEMA: &str = "ocx-resolve/1";

/// The CLI's liveness verdict. Only two reach the wire; the third exits 1.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Status {
    Live,
    AbsentProven,
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Liveness {
    pub status: Status,
    pub pid: Option<u32>,
    pub port: Option<u16>,
    /// The bind address that answered. Absent on a proven absence, because nothing answered.
    pub hostname: Option<String>,
    pub version: Option<String>,
    pub role: Option<String>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Port {
    /// The port a client should use: the live listener's, or the configured one.
    pub effective: u16,
    /// What a start would prefer.
    pub configured: u16,
}

/// Whether the CLI says a desktop takeover can be offered.
///
/// The token is the binding a later `ocx service claim` repeats back: it covers the exact
/// subject and managing-CLI observations the consent was approved against, so a claim made
/// after either moved is refused rather than recorded.
#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum Takeover {
    #[serde(rename_all = "camelCase")]
    Supported {
        protocol_version: u64,
        minimum_cli_version: String,
        token: String,
    },
    Blocked {
        reason: String,
        detail: String,
    },
}

impl Default for Takeover {
    /// An older bundled CLI carries no takeover answer at all; silence is not approval.
    fn default() -> Self {
        Self::Blocked {
            reason: "unreported".to_owned(),
            detail: "the bundled CLI did not report takeover compatibility".to_owned(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Resolved {
    pub schema: String,
    pub cli_version: String,
    pub config_home: String,
    pub port: Port,
    pub liveness: Liveness,
    /// The recorded runtime owner, already in the CLI's three answers. Absent on older
    /// documents, which read as unknown rather than as nobody owning the runtime.
    #[serde(default)]
    pub ownership: Recorded,
    #[serde(default)]
    pub takeover: Takeover,
}

impl Resolved {
    pub fn endpoint(&self) -> ProxyEndpoint {
        ProxyEndpoint {
            host: "127.0.0.1",
            port: self.port.effective,
        }
    }

    pub fn home(&self) -> PathBuf {
        PathBuf::from(&self.config_home)
    }
}

/// What the shell got back.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Resolution {
    /// The CLI produced a verdict this shell trusts.
    Answered(Box<Resolved>),
    /// It did not, for whatever reason. Never read as absence.
    Unknown(String),
}

impl Resolution {
    pub fn resolved(&self) -> Option<&Resolved> {
        match self {
            Self::Answered(resolved) => Some(resolved.as_ref()),
            Self::Unknown(_) => None,
        }
    }

    pub fn reason(&self) -> Option<&str> {
        match self {
            Self::Unknown(reason) => Some(reason),
            Self::Answered(_) => None,
        }
    }
}

/// Whether the shell may start a runtime of its own.
///
/// Proven absence and nothing else. `live` means attach to what is there, and unknown means refuse:
/// a resolution that could not be trusted must never read as "nobody is listening", which is the
/// reading that puts a second proxy next to the one already running.
pub fn may_start(resolution: &Resolution) -> bool {
    matches!(
        resolution
            .resolved()
            .map(|resolved| resolved.liveness.status),
        Some(Status::AbsentProven)
    )
}

/// What the shell may do with a listener the CLI found alive.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum LiveVerdict {
    /// Nothing is listening; this verdict does not apply.
    NotLive,
    /// It is a proxy, on an address this shell can reach. Attach as a guest.
    Attach,
    /// Something is listening and this shell cannot use it. Never a reason to start a second one.
    Unusable(String),
}

/// Whether the address the CLI reported is one this shell can reach on loopback.
///
/// The shell speaks to loopback and nothing else — that is what makes sending the management token
/// to it safe. A proxy bound to either loopback spelling, or to every interface, is reachable at
/// 127.0.0.1. One bound to the IPv6 loopback or to a specific external address is not, and
/// addressing 127.0.0.1 anyway would turn a running proxy into a health wait that times out.
pub fn loopback_reachable(hostname: Option<&str>) -> bool {
    matches!(
        hostname,
        None | Some("127.0.0.1") | Some("localhost") | Some("0.0.0.0")
    )
}

/// Read a live verdict.
///
/// Liveness answers "is something there", and core's predicate accepts a connected client's
/// listener on purpose so duplicate-start avoidance can see it. This shell needs the management
/// plane, so it has to discriminate on the role the CLI carried: a client listener serves machine
/// routes, not `/api/*`, and attaching to it would report Ready against an endpoint the dashboard
/// and the tray cannot use.
pub fn live_verdict(resolution: &Resolution) -> LiveVerdict {
    let Some(resolved) = resolution.resolved() else {
        return LiveVerdict::NotLive;
    };
    if resolved.liveness.status != Status::Live {
        return LiveVerdict::NotLive;
    }
    if resolved.liveness.role.as_deref() == Some("client") {
        return LiveVerdict::Unusable(
            "a connected client is listening on this port, not a proxy this app can manage".into(),
        );
    }
    if !loopback_reachable(resolved.liveness.hostname.as_deref()) {
        return LiveVerdict::Unusable(format!(
            "the runtime is bound to {} and this app only speaks to loopback",
            resolved
                .liveness
                .hostname
                .as_deref()
                .unwrap_or("an unknown address")
        ));
    }
    LiveVerdict::Attach
}

/// Read one resolve document, refusing anything that is not exactly one.
///
/// The CLI puts the document on stdout and its human output on stderr, so stdout is parsed whole.
/// A non-zero exit is the CLI's own refusal — including the exit 1 it uses for unknown liveness and
/// for a config it will not guess at — and is carried through rather than reinterpreted here.
pub fn read(exit_code: Option<i32>, stdout: &[u8], stderr: &[u8]) -> Resolution {
    if exit_code != Some(0) {
        let detail = String::from_utf8_lossy(stderr);
        let detail = detail.trim();
        let code = exit_code
            .map(|code| code.to_string())
            .unwrap_or_else(|| "no exit code".to_owned());
        return Resolution::Unknown(if detail.is_empty() {
            format!("the bundled CLI could not resolve the runtime (exit {code})")
        } else {
            format!("the bundled CLI could not resolve the runtime (exit {code}): {detail}")
        });
    }
    let text = String::from_utf8_lossy(stdout);
    let resolved: Resolved = match serde_json::from_str(text.trim()) {
        Ok(resolved) => resolved,
        Err(error) => {
            return Resolution::Unknown(format!(
                "the bundled CLI's resolve output could not be read ({error})"
            ))
        }
    };
    if resolved.schema != SCHEMA {
        return Resolution::Unknown(format!(
            "the bundled CLI answered with schema {} and this app understands {SCHEMA}",
            resolved.schema
        ));
    }
    Resolution::Answered(Box::new(resolved))
}

/// Ask the bundled CLI, under the caller's deadline.
pub async fn run(app: &AppHandle, deadline: Instant) -> Resolution {
    let command = match app.shell().sidecar("ocx") {
        Ok(command) => command.args(["resolve", "--json"]),
        Err(error) => {
            return Resolution::Unknown(format!("the bundled CLI could not be started ({error})"))
        }
    };
    match timeout_at(deadline, command.output()).await {
        Ok(Ok(output)) => read(output.status.code(), &output.stdout, &output.stderr),
        Ok(Err(error)) => {
            Resolution::Unknown(format!("the bundled CLI could not be run ({error})"))
        }
        Err(_) => Resolution::Unknown(
            "the bundled CLI did not answer before the startup deadline".to_owned(),
        ),
    }
}

#[cfg(test)]
mod tests {
    use super::{
        live_verdict, loopback_reachable, may_start, read, LiveVerdict, Resolution, Status,
        Takeover, SCHEMA,
    };
    use crate::ownership::{Owner, Recorded};

    const LIVE: &str = r#"{"schema":"ocx-resolve/1","cliVersion":"2.61.0","configHome":"/h",
        "port":{"effective":10100,"configured":10100,"source":"runtime-record"},
        "liveness":{"status":"live","pid":42,"port":10100,"source":"runtime-record","version":"2.61.0"}}"#;
    const ABSENT: &str = r#"{"schema":"ocx-resolve/1","cliVersion":"2.61.0","configHome":"/h",
        "port":{"effective":10100,"configured":10100,"source":"config"},
        "liveness":{"status":"absent-proven","pid":null,"port":null,"source":null}}"#;

    #[test]
    fn a_live_verdict_is_read_whole() {
        let resolution = read(Some(0), LIVE.as_bytes(), b"");
        let resolved = resolution.resolved().expect("a document");
        assert_eq!(resolved.schema, SCHEMA);
        assert_eq!(resolved.liveness.status, Status::Live);
        assert_eq!(resolved.liveness.pid, Some(42));
        assert_eq!(resolved.endpoint().port, 10100);
        assert_eq!(resolved.home().display().to_string(), "/h");
        assert_eq!(live_verdict(&resolution), LiveVerdict::Attach);
        assert!(!may_start(&resolution));
    }

    #[test]
    fn only_a_proven_absence_authorises_a_start() {
        let resolution = read(Some(0), ABSENT.as_bytes(), b"");
        assert_eq!(
            resolution.resolved().map(|r| r.liveness.status),
            Some(Status::AbsentProven)
        );
        assert!(may_start(&resolution));
        assert_eq!(live_verdict(&resolution), LiveVerdict::NotLive);
    }

    #[test]
    fn a_connected_client_is_live_but_not_a_runtime_to_attach_to() {
        let client = LIVE.replace(
            r#""version":"2.61.0""#,
            r#""version":"2.61.0","role":"client""#,
        );
        let resolution = read(Some(0), client.as_bytes(), b"");
        assert!(matches!(
            live_verdict(&resolution),
            LiveVerdict::Unusable(_)
        ));
        // Live and unusable is still live: it is never a reason to start a second one.
        assert!(!may_start(&resolution));
    }

    #[test]
    fn only_a_loopback_bind_is_addressed_as_loopback() {
        for reachable in [None, Some("127.0.0.1"), Some("localhost"), Some("0.0.0.0")] {
            assert!(loopback_reachable(reachable), "{reachable:?}");
        }
        for elsewhere in [Some("::1"), Some("192.168.1.10"), Some("example.internal")] {
            assert!(!loopback_reachable(elsewhere), "{elsewhere:?}");
        }
        let bound = LIVE.replace(r#""pid":42"#, r#""pid":42,"hostname":"::1""#);
        let resolution = read(Some(0), bound.as_bytes(), b"");
        assert!(matches!(
            live_verdict(&resolution),
            LiveVerdict::Unusable(_)
        ));
        assert!(!may_start(&resolution));
    }

    #[test]
    fn the_clis_own_refusal_is_unknown_and_never_authorises_a_start() {
        // Exit 1 is what the CLI uses for unknown liveness and for a config it will not guess at.
        let resolution = read(Some(1), b"", b"resolve: liveness is unknown");
        assert!(matches!(resolution, Resolution::Unknown(_)));
        assert!(resolution.reason().unwrap().contains("liveness is unknown"));
        assert!(!may_start(&resolution));
        assert_eq!(live_verdict(&resolution), LiveVerdict::NotLive);
    }

    #[test]
    fn everything_that_can_go_wrong_here_folds_into_unknown() {
        for (code, out) in [
            (Some(64), &b""[..]),
            (None, &b""[..]),
            (Some(0), &b"not json"[..]),
            (Some(0), &b"{}"[..]),
        ] {
            let resolution = read(code, out, b"");
            assert!(matches!(resolution, Resolution::Unknown(_)), "{code:?}");
            assert!(!may_start(&resolution));
        }
    }

    #[test]
    fn ownership_and_takeover_answers_are_read_whole() {
        let document = format!(
            "{}{}}}",
            LIVE.strip_suffix('}').unwrap(),
            r#","ownership":{"kind":"owned","ownership":{"owner":"cli","installId":"npm-1","consentGeneration":2},"revision":9},"takeover":{"kind":"supported","protocolVersion":1,"minimumCliVersion":"2.61.0","token":"abc"}"#
        );
        let resolution = read(Some(0), document.as_bytes(), b"");
        let resolved = match resolution.resolved() {
            Some(resolved) => resolved.clone(),
            None => panic!("{}", resolution.reason().unwrap()),
        };
        assert_eq!(
            resolved.ownership,
            Recorded::Owned {
                ownership: crate::ownership::Claim {
                    owner: Owner::Cli,
                    install_id: "npm-1".to_owned(),
                    consent_generation: 2,
                },
                revision: 9,
            }
        );
        assert!(matches!(
            resolved.takeover,
            Takeover::Supported { ref token, .. } if token == "abc"
        ));
    }

    #[test]
    fn a_missing_ownership_or_takeover_answer_is_not_consent() {
        // Older bundled CLIs carry neither field; silence must read unknown/blocked, never
        // "nobody owns it" or "takeover supported".
        let resolved = read(Some(0), LIVE.as_bytes(), b"")
            .resolved()
            .expect("a document")
            .clone();
        assert!(matches!(resolved.ownership, Recorded::Unknown { .. }));
        assert!(matches!(resolved.takeover, Takeover::Blocked { .. }));
        assert_eq!(resolved.takeover, Takeover::default());
    }

    #[test]
    fn a_blocked_takeover_carries_its_reason() {
        let document = format!(
            "{}{}}}",
            LIVE.strip_suffix('}').unwrap(),
            r#","ownership":{"kind":"none","revision":0},"takeover":{"kind":"blocked","reason":"managing-cli-unsupported","detail":"path uses 2.59.0","minimumCliVersion":"2.61.0"}"#
        );
        let resolved = read(Some(0), document.as_bytes(), b"")
            .resolved()
            .expect("a document")
            .clone();
        assert_eq!(resolved.ownership, Recorded::None { revision: 0 });
        assert_eq!(
            resolved.takeover,
            Takeover::Blocked {
                reason: "managing-cli-unsupported".to_owned(),
                detail: "path uses 2.59.0".to_owned(),
            }
        );
    }

    #[test]
    fn a_schema_this_app_does_not_know_is_unknown() {
        let future = LIVE.replace("ocx-resolve/1", "ocx-resolve/2");
        let resolution = read(Some(0), future.as_bytes(), b"");
        assert!(matches!(resolution, Resolution::Unknown(_)));
        assert!(resolution.reason().unwrap().contains("ocx-resolve/2"));
        assert!(!may_start(&resolution));
    }
}
